presidential-speeches-rag/presidential-speeches-rag.ipynb

```diff
@@ -0,0 +1,657 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "c6057e9f-874e-4d7a-9f3c-e411a9acbb2e",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "3637"
+      ]
+     },
+     "execution_count": 5,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "model_id = \"mistralai/Mixtral-8x7B-v0.1\"\n",
+    "tokenizer = AutoTokenizer.from_pretrained(model_id)\n",
+    "#tokenizer = tiktoken.get_encoding('p50k_base')\n",
+    "\n",
+    "# create the length function\n",
+    "def token_len(text):\n",
+    "    tokens = tokenizer.encode(\n",
+    "        text\n",
+    "    )\n",
+    "    return len(tokens)\n",
+    "\n",
+    "token_len(garfield_inaugural)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "71de3b6c-e89b-4446-a585-32cf10a1cd8e",
+   "metadata": {},
+   "source": [
+    "Next, we'll split the text into chunks using LangChain's `TokenTextSplitter` function. In this example we will set the maximum tokens in a chunk to be 450, with a 20 token overlap to reduce the chances that a sentence or concept will be split into different chunks.\n",
+    "\n",
+    "Note that LangChain uses OpenAI's `tiktoken` tokenizer, so our Mixtral tokenizer will count tokens a bit differently - when adjusting for this, our chunk sizes will be around 500 tokens."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "20ba719b-9a03-437a-a665-a0bde9ec24cf",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "488\n",
+      "490\n",
+      "492\n",
+      "488\n",
+      "481\n",
+      "478\n",
+      "491\n",
+      "388\n"
+     ]
+    }
+   ],
+   "source": [
+    "text_splitter = TokenTextSplitter(\n",
+    "    chunk_size=450, # 500 tokens is the max\n",
+    "    chunk_overlap=20 # Overlap of N tokens between chunks (to reduce chance of cutting out relevant connected text like middle of sentence)\n",
+    ")\n",
+    "\n",
+    "chunks = text_splitter.split_text(garfield_inaugural)\n",
+    "\n",
+    "for chunk in chunks:\n",
+    "    print(token_len(chunk))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ce723eea-7e69-48c1-8452-957709d117db",
+   "metadata": {},
+   "source": [
+    "Next, we will embed each chunk into a semantic vector space using the all-MiniLM-L6-v2 model, through LangChain's implementation of Sentence Transformers from [HuggingFace](https://huggingface.co/). Note that each embedding has a length of 384."
```

Comment:
Not super important, but maybe a specific link to the implementation instead of just a link to HF?